{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPWVo6lhUmWYJAS+qTCirlT",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/alex-tianhuang/idrfeatlib/blob/main/notebooks/Featurize.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Welcome!\n",
        "========\n",
        "\n",
        "Welcome to featurize.ipynb, a quick and easy colab notebook for computing some sequence features over a file of IDRs.\n",
        "\n",
        "To use this notebook for the first time, hover your mouse over the top left corner (around where the play buttons or cell numbers are) of each cell to run it. Just run each of the three cells in order and respond to the prompts below.\n",
        "\n",
        "To reuse this notebook, keep running the third cell below and responding to the prompts differently or uploading different files.\n",
        "\n",
        "What sequence features are there?\n",
        "=================================\n",
        "\n",
        "At its simplest, this notebook takes in sequences (a FASTA file), and outputs sequence features in the form of a very wide csv.\n",
        "\n",
        "A full table of all the features and their interpretations will come soon. However, as a birds-eye view guide, generally speaking, the outputs start like:\n",
        "```\n",
        "ProteinID,AA_A,AA_C,...,AA_Y,...\n",
        "Protein1,0.039,0.01,...,0.02,...\n",
        "Protein2,0.045,0.02,...,0.02,...\n",
        "...\n",
        "```\n",
        "\n",
        "which indicates `Protein1` is made of 3.9% alanines, 1% cysteines, ..., and 2% tyrosines.\n",
        "\n",
        "Then there are many columns (as of June 24th, 2025 there are 89) motif or repeat related columns.\n",
        "\n",
        "Some columns like `D_repeats` are simple to understand and correspond to the number of amino acids spanned by Ds that are followed immediately followed by another D (D repeats).\n",
        "\n",
        "Some of the columns like `DOC_PP1_RVXF_1` or `LIG_PCNA_PIPBox1` come from names of motifs in the Eukaryotic Linear Motif (ELM) resource.\n",
        "These are integer valued counts of motifs or the number of amino acids spanned by those motifs, and you can read more about them by searching them up in the ELM.\n",
        "\n",
        "Then there are some columns that correspond to the percentage of sequence spanned by groups of amino acids like:\n",
        "\n",
        "```\n",
        "ProteinID,...,acidic,basic,fcr,...\n",
        "Protein1,...,0.05,0.14,0.27,...\n",
        "Protein2,...,0.06,0.12,0.30,...\n",
        "...\n",
        "```\n",
        "\n",
        "There are a couple that correspond to residue spacing statistics like `custom_kappa`, `arospacing`, `SCD`...\n",
        "\n",
        "And some specialized sequence features like `isoelectric_point`, `net_charge`, and so on."
      ],
      "metadata": {
        "id": "fhhwrXblt8Yh"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Setup environment by downloading and installing the idrfeatlib repo.\n",
        "#\n",
        "# You only need to run this once.\n",
        "!file idrfeatlib/ >/dev/null && rm -rf idrfeatlib >/dev/null\n",
        "!git clone https://github.com/alex-tianhuang/idrfeatlib --quiet\n",
        "!pip install idrfeatlib/"
      ],
      "metadata": {
        "id": "ZNBlM51RZiZT"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "c-7itQrJqTI-"
      },
      "outputs": [],
      "source": [
        "# Prepare/define the featurization scripts.\n",
        "\n",
        "def main(args):\n",
        "    \"\"\"\n",
        "    Taken from idrfeatlib/featurize.py (25-06-24) @tianh.\n",
        "    \"\"\"\n",
        "    from idrfeatlib.utils import read_fasta, read_regions_csv, iter_nested\n",
        "    from idrfeatlib import FeatureVector\n",
        "    from idrfeatlib.featurizer import Featurizer, compile_featurizer\n",
        "    from idrfeatlib.native import compile_native_featurizer\n",
        "    import json\n",
        "    import tqdm\n",
        "    import sys\n",
        "    if args.feature_file:\n",
        "        with open(args.feature_file, \"r\") as file:\n",
        "            config = json.load(file)\n",
        "        featurizer, errors = compile_featurizer(config)\n",
        "    else:\n",
        "        featurizer, errors = compile_native_featurizer()\n",
        "    for featname, error in errors.items():\n",
        "        print(\"error compiling `%s`: %s\" % (featname, error), file=sys.stderr)\n",
        "    featurizer = Featurizer(featurizer)\n",
        "    if args.rtf:\n",
        "        fa = read_fasta_rtf(args.input_sequences)\n",
        "    else:\n",
        "        fa = read_fasta(args.input_sequences)\n",
        "    if args.input_regions is None:\n",
        "        seqs = tqdm.tqdm(fa, total=len(fa), desc=\"featurizing sequences\")\n",
        "        feature_vectors, errors = featurizer.featurize_to_matrices(seqs)\n",
        "        for proteinid, error in errors.items():\n",
        "            print(\"error for `%s`: %s\" % (proteinid, error), file=sys.stderr)\n",
        "        FeatureVector.dump(list(feature_vectors.items()), args.output_file, \"ProteinID\")\n",
        "    else:\n",
        "        fa = dict(fa)\n",
        "        regions = read_regions_csv(args.input_regions)\n",
        "        seqs = []\n",
        "        for protid, regionid, (start, stop) in iter_nested(regions, 2):\n",
        "            if (entry := fa.get(protid)) is None:\n",
        "                continue\n",
        "            whole_seq = entry\n",
        "            assert isinstance(whole_seq, str), type(whole_seq)\n",
        "            seq = whole_seq[start:stop]\n",
        "            if len(seq) != stop - start:\n",
        "                print(\"invalid region `%s` for protein `%s` (start=%d,stop=%d,seqlen=%s)\" % (regionid, protid, start, stop, len(whole_seq)))\n",
        "                continue\n",
        "            seqs.append(((protid, regionid), seq))\n",
        "        feature_vectors, errors = featurizer.featurize_to_matrices(tqdm.tqdm(seqs, desc=\"featurizing sequences\"))\n",
        "        for (protid, regionid), error in errors.items():\n",
        "            print(\"error for (protid=%s, regionid=%s): %s\" % (protid, regionid, error), file=sys.stderr)\n",
        "        FeatureVector.dump(list(feature_vectors.items()), args.output_file, (\"ProteinID\", \"RegionID\"))\n",
        "\n",
        "def display_csv(output_name):\n",
        "    \"\"\"\n",
        "    Show the table in the notebook.\n",
        "\n",
        "    I assume colab will forever keep pandas as available by default.\n",
        "    \"\"\"\n",
        "    from IPython.display import display\n",
        "    import pandas as pd\n",
        "\n",
        "    df = pd.read_csv(output_name)\n",
        "\n",
        "    print()\n",
        "    print(\"Showing output below\")\n",
        "    print(\"--------------------\")\n",
        "    print()\n",
        "    display(df)\n",
        "    print()\n",
        "\n",
        "def run_colab_wrapper(output_name, rtf=False):\n",
        "    \"\"\"\n",
        "    My `main` function expects an argparse.Namespace and I'm not changing it\n",
        "    so most of this function is weirdly constructing this one Namespace object\n",
        "    \"\"\"\n",
        "    from google.colab import files\n",
        "    import os\n",
        "    import argparse\n",
        "\n",
        "    # The aforementioned Namespace object, starting as an empty args\n",
        "    args = argparse.ArgumentParser().parse_args([])\n",
        "\n",
        "    args.input_sequences = 'input_sequences.fasta'\n",
        "    goto_upload = True\n",
        "    if os.path.exists(args.input_sequences):\n",
        "        choice = input(f\"The file {args.input_sequences} already exists. Would you like to overwrite it? (y/n)\")\n",
        "        if choice.lower() != 'y':\n",
        "            goto_upload = False\n",
        "    if goto_upload:\n",
        "        files.upload_file(args.input_sequences)\n",
        "\n",
        "    choice = input(\"Would you like to upload a file containing region boundaries? (y/n)\")\n",
        "    if choice.lower() == 'y':\n",
        "        args.input_regions = 'input_regions.csv'\n",
        "        files.upload_file(args.input_regions)\n",
        "    else:\n",
        "        args.input_regions = None\n",
        "\n",
        "    args.feature_file = 'feature_config.json'\n",
        "    goto_upload = True\n",
        "    if os.path.exists(args.feature_file):\n",
        "        choice = input(f\"The file {args.feature_file} already exists. Would you like to overwrite it? (y/n)\")\n",
        "        if choice.lower() != 'y':\n",
        "            goto_upload = False\n",
        "            print(f\"Ignoring {args.feature_file}\")\n",
        "    else:\n",
        "        choice = input(\"Would you like to upload a file containing feature configuration? (y/n)\")\n",
        "        if choice.lower() != 'y':\n",
        "            goto_upload = False\n",
        "    if goto_upload:\n",
        "        files.upload_file(args.feature_file)\n",
        "    else:\n",
        "        args.feature_file = None\n",
        "\n",
        "    abort = False\n",
        "    args.output_file = output_name\n",
        "    if os.path.exists(args.output_file):\n",
        "        choice = input(f\"The file {args.output_file} already exists. Would you like to overwrite it? (y/n)\")\n",
        "        if choice.lower() != 'y':\n",
        "            abort = True\n",
        "    if abort:\n",
        "        print(f\"Aborting because user wants to do something with {args.output_file}.\")\n",
        "        return\n",
        "\n",
        "    args.rtf = rtf\n",
        "\n",
        "    main(args)\n",
        "    display_csv(args.output_file)\n",
        "    print(f\"Downloading output file to {args.output_file}\")\n",
        "    files.download(args.output_file)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# This cell will:\n",
        "#\n",
        "# 1. Ask for input files:\n",
        "#    - (required) a FASTA file of sequences\n",
        "#    - (optional) a CSV file of region boundaries, see extra section below\n",
        "#    - (optional) a JSON file for defining custom features, see extra section below\n",
        "# 2. Featurize sequences from the FASTA file\n",
        "# 3. Ask to download the output file, called `output_features.csv`\n",
        "#\n",
        "# Run this cell after running the above cells as many times as you would like.\n",
        "\n",
        "run_colab_wrapper(\"output_features.csv\")\n",
        "print(\"Done!\")\n"
      ],
      "metadata": {
        "id": "BS5xozdVZgmj"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Troubleshooting\n",
        "===============\n",
        "\n",
        "FASTA parsing errors\n",
        "--------------------\n",
        "\n",
        "When you see many warnings that look like:\n",
        "\n",
        "    /usr/local/lib/python3.11/dist-packages/idrfeatlib/utils.py:24: UserWarning: non-empty line before first fasta header: ...\n",
        "\n",
        "Typically the script will run to completion and print `Done!` but the file downloaded is a CSV with no data, only a header.\n",
        "\n",
        "This typically means that you have not selected something that looks like a FASTA file, with `>` header lines. If you're still getting that error but it looks like a a FASTA file, see the `RTF files` section below.\n",
        "\n",
        "RTF files\n",
        "---------\n",
        "\n",
        "Sometimes, when viewing a fasta.rtf file in a rich text editor (like TextEdit on MacOS), it will appear as if there are no characters before the first `>` but there will be errors like:\n",
        "\n",
        "    /usr/local/lib/python3.11/dist-packages/idrfeatlib/utils.py:24: UserWarning: non-empty line before first fasta header: {\\rtf1\\ansi\\ansicpg1252\\cocoartf2513\n",
        "\n",
        "Again, the script runs to completion and prints `Done!` but the file downloaded is a CSV with no data, only a header.\n",
        "\n",
        "This is likely because there are invisible characters that help format the text but are messing up the FASTA parser. To solve this issue, you can run the two code cells below:"
      ],
      "metadata": {
        "id": "-qTOaHY42nfD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Install and prepare a fasta.rtf parsing script for use in the cell below.\n",
        "!pip install striprtf\n",
        "from striprtf.striprtf import rtf_to_text\n",
        "import warnings\n",
        "def read_fasta_rtf(path):\n",
        "    \"\"\"Same as idrfeatlib.read_fasta, except with rtf_to_text before it.\"\"\"\n",
        "    with open(path, \"r\") as file:\n",
        "        text = file.read()\n",
        "    content = rtf_to_text(text)\n",
        "    fasta_list = []\n",
        "    for line in content.splitlines():\n",
        "        line = line.strip()\n",
        "        if line.startswith(\">\"):\n",
        "            header = line[1:]\n",
        "            fasta_list.append([header, []])\n",
        "        elif fasta_list:\n",
        "            fasta_list[-1][1].append(line)\n",
        "        elif line:\n",
        "            warnings.warn(\"Ignoring non-empty line before first fasta header: {}\".format(line))\n",
        "    for index, (header, seqlines) in enumerate(fasta_list):\n",
        "        fasta_list[index] = (header, \"\".join(seqlines))\n",
        "    return fasta_list\n"
      ],
      "metadata": {
        "id": "F9RIGMXv7FnF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# This cell will:\n",
        "#\n",
        "# 1. Ask for input files:\n",
        "#    - (required) a FASTA.RTF file of sequences\n",
        "#    - (optional) a CSV file of region boundaries, see extra section below\n",
        "#    - (optional) a JSON file for defining custom features, see extra section below\n",
        "# 2. Featurize sequences from the FASTA file\n",
        "# 3. Ask to download the output file, called `output_features.csv`\n",
        "#\n",
        "# Run this cell after running the above cells as many times as you would like.\n",
        "\n",
        "run_colab_wrapper(\"output_features.csv\", rtf=True)\n",
        "print(\"Done!\")"
      ],
      "metadata": {
        "id": "Zhd2bND88g9r"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Extra\n",
        "=====\n",
        "The absolutely essential bits are above, but below are some extra tips.\n",
        "\n",
        "\n",
        "Uploading regions\n",
        "-----------------\n",
        "If you have a fasta file of full protein sequences and a csv full of named protein regions and their start/stop boundaries, like the following:\n",
        "```\n",
        "ProteinID,RegionID,Start,Stop\n",
        "DDX3X,N-IDR,0,142\n",
        "protein_x,region3,24,72\n",
        "...\n",
        "```\n",
        "Then upload this file when asked about a region boundaries csv and the notebook will automatically cut out regions for you.\n",
        "\n",
        "Using custom features\n",
        "---------------------\n",
        "Follow this section to featurize using your own features.\n",
        "\n",
        "Suppose you have a script uploaded @ `FOO.py`:\n",
        "\n",
        "    def bar(sequence, secret_param):\n",
        "        return len(sequence) + secret_param\n",
        "\n",
        "To turn this into a sequence feature, make a json object that looks like this:\n",
        "\n",
        "    {\n",
        "        \"features\": {\n",
        "            \"my_feature\": {\n",
        "                \"compute\": \"custom\",\n",
        "                \"libpath\": \"FOO.py\",\n",
        "                \"funcname\": \"bar\",\n",
        "                \"kwargs\": {\n",
        "                    \"secret_param\": 42\n",
        "                }\n",
        "            }\n",
        "        }\n",
        "    }\n",
        "\n",
        "Upload it into the file `feature_config.json`, so that when prompted about the features json file, you can upload this file and compute `len(seq) + 42` for arbitrary sequences (for example).\n",
        "\n",
        "If you have multiple functions, you can use multiple names this way like:\n",
        "\n",
        "    {\n",
        "        \"features\": {\n",
        "            \"feature_a\": ...,\n",
        "            \"feature_b\": ...,\n",
        "            ...\n",
        "        }\n",
        "    }\n",
        "\n",
        "Admittedly the use of this feature format to run pre-existing python functions is kind of cumbersome. However if you get used to doing it this way there's a chance you'll find it more useful when integrating with other tools (not currently available but maybe someday) that use this feature specification format."
      ],
      "metadata": {
        "id": "aSNQtUPyyqyf"
      }
    }
  ]
}